### DISCUSSION TITLE
kernel unable to find an executable

### DISCUSSION THREAD

**Discussion by amit-chandak-unskript** [jupyter/enterprise_gateway#1003] [Q&A] [answered]
### Environment

- Enterprise Gateway Version
elyra/enterprise-gateway:2.5.0, with ipykernel upgraded to 6.3.1

- Platform: Kubernetes
- Jupyter version
```
(base) jovyan@f24df960513e:~$ jupyter --version
jupyter core     : 4.7.1
jupyter-notebook : 6.4.3
qtconsole        : not installed
ipython          : 7.26.0
ipykernel        : 6.1.0
jupyter client   : 6.1.12
jupyter lab      : 3.1.6
nbconvert        : 6.1.0
ipywidgets       : not installed
nbformat         : 5.1.3
traitlets        : 5.0.5
```

I need to run a notebook which requires running an executable (inside python code). This executable is aws-iam-authenticator. This is used to authenticate with AWS to talk to AWS EKS.

We use our custom kernel, based off **elyra/kernel-py:2.5.0** and in that i package the executable, aws-iam-authenticator, and put it in /usr/local/bin, as confirmed on the kernel docker
```
jovyan@edf953e24647:~$ ls -l /usr/local/bin/
total 34492
-rwxr-xr-x 1 root   users 35290752 Jul 14 22:56 aws-iam-authenticator
-rwxr-xr-x 1 jovyan users     3814 Feb 16  2021 bootstrap-kernel.sh
-rwxr-xr-x 1 root   root       965 Jun 18  2020 fix-permissions
drwxr-xr-x 1 jovyan users     4096 Feb 16  2021 kernel-launchers
-rwxr-xr-x 1 root   root       534 Jun 18  2020 start-notebook.sh
-rwxr-xr-x 1 root   root      5648 Jun 18  2020 start.sh
-rwxr-xr-x 1 root   root      1197 Jun 18  2020 start-singleuser.sh
jovyan@edf953e24647:~$
```

However, i keep getting this error, when i run the python code which requires running this executable:

```
[E 2021-09-09 06:37:01,054.054 root] [Errno 2] No such file or directory: 'aws-iam-authenticator': 'aws-iam-authenticator'
```

Couple of questions:

1. Is /usr/local/bin the right place to park the executable? 
2. I was under the impression that the python code, which the kernel runs, runs under **jovyan**? If thats right, why does the error above mentions **root**?
3. Is the owner of the executable being root ok? or does it need to be jovyan? When i get into the kernel using bash, its able to run the executable, so its definitely in the path and with right permissions.

What else could explain this issue? Really appreciate your help. @kevin-bates, do you have any ideas?

Thanks
Amit

**Comment by amit-chandak-unskript** [comment, marked as answer]
I found the issue, somehow the kernel version was not having the executable packaged in (some version issue). After i fixed the version issue, it is working. Sorry for the trouble.

**Comment by kevin-bates** [comment]
Cool - thanks for the update.  Here's a follow-up to your questions (for completeness):

> Is /usr/local/bin the right place to park the executable?

Yes.  That should work and is in the container's PATH.

> I was under the impression that the python code, which the kernel runs, runs under jovyan? If thats right, why does the error above mentions root?

Hmm, not sure. I went into the elyra/kernel-py:2.5.1 container and found that my id was 'jovyan' and our dockerfiles switch back to jovyan via `USER jovyan`, so would expect the effective user to not be root.  I see that `/usr/local/bin` is only writable by root, so perhaps the update with the AWS executable didn't revert `USER`.

> Is the owner of the executable being root ok? or does it need to be jovyan? When i get into the kernel using bash, its able to run the executable, so its definitely in the path and with right permissions.

That should be prevented and, when using UID and GID values, we try our best to prevent 0 (zeros).
